### ISSUE TITLE
false positive for PERF203 with try-except-else

### ISSUE THREAD

**Issue by trim21** [astral-sh/ruff#11602]
ruff 0.4.6 raise false positive for try-except in a look with `else`

```python
import io
import unicodedata


def __clean_rss_invalid_char(s: str):
    with io.StringIO() as f:
        for c in s:
            try:
                unicodedata.name(c)
            except ValueError:
                continue
            else:
                f.write(c)

        return f.getvalue()
```

previous issue: https://github.com/astral-sh/ruff/issues/6535

**Comment by trim21** [comment]
Oh, sorry. looks like I'm actually using ruff 0.4.5.

**Comment by MichaReiser** [comment]
Looking at the code, the reason why the above is flagged and wasn't addressed by #6535 is that the rule only checks for `continue` and `break` statements inside the `try` body but not in the except handlers

https://github.com/astral-sh/ruff/blob/f9d818967075df1a1f4d15760c0244f941d77e53/crates/ruff_linter/src/rules/perflint/rules/try_except_in_loop.rs#L106-L108

**Comment by MichaReiser** [comment]
I still think the rule should flag the above pattern because whether there's a better way to write the above loop depends on the `unicodedata` API. For example, I think the following could work:

```python
for c in s:
    try:
        name = unicodedata.name("a", None)

        if name is None:
            continue
        else:
            f.write(c)
```

**Comment by trim21** [comment]
> I still think the rule should flag the above pattern because whether there's a better way to write the above loop depends on the `unicodedata` API. For example, I think the following could work:
> 
> ```python
> for c in s:
>       name = unicodedata.name("a", None)
> 
>       if name is None:
>           continue
>       else:
>           f.write(c)
> ```
> 
> which eliminates the need for try

That's not a valid solution, ruff still report error for this:

```
def _is_validate_character(c: str): ...


def __clean_rss_invalid_char(s: str):
    for c in s:
        try:
            _is_validate_character(c)
        except ValueError:
            continue
        else:
            print(c)

```

**Comment by trim21** [comment]
also it doesn't flag this, even though they are equal:

```python
import requests


def check_url_live(urls: str):
    for url in urls:
        try:
            requests.get(url, timeout=5)
        except ConnectionRefusedError:
            continue

        print(url)
```

**Comment by MichaReiser** [comment]
> because ruff still flag this, and I don't think there is a better way to write this:

Yeah, there's probably not a better way to write this and using a noqa comment here seems appropriate. 

> if you think it should flag this, then there is another problem: it doesn't flag this:

I would have to double check the upstream rule to understand if this is a bug but the rule currently only flags cases where the `try` statement is the only statement in the loop body.

**Comment by MichaReiser** [comment]
Going through the history. Ignoring multi-statement loops is intentional, see https://github.com/astral-sh/ruff/pull/6145

Using noqa comment seems the right solution to deal with cases where using `try...except` is the only solution because the API doesn't provide an alternative way of testing the input. Making this decision isn't something that ruff can do (even with better inference).

**Comment by trim21** [comment]
> Going through the history. Ignoring multi-statement loops is intentional, see #6145
> 
> Using noqa comment seems the right solution to deal with cases where using `try...except` is the only solution because the API doesn't provide an alternative way of testing the input. Making this decision isn't something that ruff can do (even with better inference).

I think we are actully talking 2 topic here... 

Ignoring multi-statement loops is fine to me, and this issue is not about it should or should not ignore multi-statement loops. the "falst positive"  is, it didn't ignore single `try-except-else` statments like multi-statement loops
